### PR TITLE
#2554 Search page re-renders on entering input 

### LINKED ISSUE #2554: Search page re-renders on entering input
**What happened**:
When entering any text inside the search bar, the search page re-renders and focus is removed from the search bar.

**What should have happened**:
When entering any text, it should keep focus so that you can continue entering your search text. This is a render issue, most likely a variable/function introduced in one of our recent PRs that caused this issue.

The issue is caused by the removal of the `autoFocus` property in [this recent commit](https://github.com/Seneca-CDOT/telescope/commit/d8a7f3330d670ec89ca499d69d94055133f46feb). Without that property, upon entering a character, our `onChange` event handler will cause a useEffect re-render from our SearchProvider, thus, the search page itself will be rendered again and the search input text field loses focus.

### PR DESCRIPTION
**Pull request by drew5494**
This PR closes #2554 

## Type of Change

- [x] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

When entering any text into search, it should keep focus so that you can continue entering your search text. The <code>autofocus</code> attribute is not recommended by ESLint, unless the page is mainly used for searching, as it is in this case. See [The accessibility of HTML 5 autofocus](https://brucelawson.co.uk/2009/the-accessibility-of-html-5-autofocus/).

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/api/status/env.local [hunk 1]
```diff
@@ -1,3 +1,7 @@
 PLANET_PORT=1111
```
Comment (HyperTHD) [inline src/api/status/env.local:1]: This file is unrelated to PR's intended fix, remove this.

File: src/api/status/src/server.js [hunk 2]
```diff
@@ -44,4 +44,4 @@ if (process.env.PATH_PREFIX) {
 }
 
 const port = parseInt(process.env.STATUS_PORT || 1111, 10);
```
Comment (HyperTHD) [inline src/api/status/src/server.js:46]: Same with this, remove this file

File: src/web/src/components/SearchInput/AuthorSearchInput.tsx [hunk 3]
```diff
@@ -33,6 +33,8 @@ const AuthorSearchInput = () => {
   return (
     <>
       <input
+        // eslint-disable-next-line jsx-a11y/no-autofocus
```
Comment (HyperTHD) [inline src/web/src/components/SearchInput/AuthorSearchInput.tsx:36]: We removed the `autoFocus` property specifically because of the eslint fix that came in. The solution for this will have to come from a different source rather than adding the autoFocus prop back in. Remove this comment and the autoFocus prop from here.

File: src/web/src/components/SearchInput/PostSearchInput.tsx [hunk 4]
```diff
@@ -35,6 +35,8 @@ const PostSearchInput = () => {
   return (
     <>
       <input
+        // eslint-disable-next-line jsx-a11y/no-autofocus
```
Comment (HyperTHD) [inline src/web/src/components/SearchInput/PostSearchInput.tsx:38]: Same thing here, remove the comment and the autoFocus prop

**Review by HyperTHD — CHANGES_REQUESTED** [review]
Alright, lots of work gonna be needed to fix this. Aside from the files being added that aren't supposed to be here, the fix you added is one we've decided to not add back anymore. You can read [this link](https://stackoverflow.com/questions/2180645/is-automatically-assigning-focus-bad-for-accessibility) to see why it's not a good idea.

**Comment by drew5494** [comment]
Alright next time close the issue first. 

> Alright, lots of work gonna be needed to fix this. Aside from the files being added that aren't supposed to be here, the fix you added is one we've decided to not add back anymore. You can read [this link](https://stackoverflow.com/questions/2180645/is-automatically-assigning-focus-bad-for-accessibility) to see why it's not a good idea.

As stated it may be acceptable for pages used for searching: "It's clear why your focus winds up in the Google search box so that doesn't bother me"
Are you still looking for an alternative to autofocus ?

**Comment by humphd** [comment]
@drew5494 we recently had to move away from using `autoFocus`, so it would be good to find a way to accomplish this that doesn't rely on it.  However, I'd have to dig up the PR that did it (it might have been `eslint` related)?

It's interesting to see that Google uses it:

```html
<input class="gLFyf gsfi" jsaction="paste:puy29d;" maxlength="2048" name="q" type="text" aria-autocomplete="both" aria-haspopup="false" autocapitalize="none" autocomplete="off" autocorrect="off" autofocus="" role="combobox" spellcheck="false" title="Search" value="" aria-label="Search" data-ved="0ahUKEwjz1dCF7sr0AhXbG80KHc4SAIAQ39UDCAY">
```

So maybe we can make eslint happy *and* do this?  Not sure.

**Comment by DukeManh** [comment]
This is introduced in my eslint PR that removed `autofocus` from input because of [Eslint's accessibility recommendations](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/no-autofocus.md). 

I have investigated a bit, the problem is here in `searchInput`:
https://github.com/Seneca-CDOT/telescope/blob/78f3324cf79f253db17a33f832e99a6e5072a4c7/src/web/src/components/SearchInput/SearchInput.tsx#L8

We have a `useSearchValue` hook to get the search `filter` and `value`, when the input changes, all components that subscribe to the hook re-renders, hence`searchInput` rerenders and returns a new component every time, either `AuthorInput` or `PostsInput`, which makes the input lose focus.

If we move the conditional into a JSX expression to make sure it returns the same component, the problem should go away.

```js
  return <>{filter === 'author' ? <AuthorSearchInput /> : <PostSearchInput />}</>;
```

File: src/web/src/components/SearchInput/AuthorSearchInput.tsx [hunk 5]
```diff
@@ -33,6 +33,8 @@ const AuthorSearchInput = () => {
...
+        autoFocus
         className={classes.input}
         list="search-suggestions"
```
Comment (DukeManh) [inline src/web/src/components/SearchInput/AuthorSearchInput.tsx:39]: While you are, can you remove this `list` attribute and the `data-list` field below? They're used for input suggestions but we don't have any.

**Review by DukeManh — CHANGES_REQUESTED** [review]

File: src/web/src/components/SearchInput/AuthorSearchInput.tsx (hunk 5 above)
Comment (drew5494) [reply src/web/src/components/SearchInput/AuthorSearchInput.tsx:39]: sure @DukeManh

File: src/api/status/env.local [hunk 6]
```diff
@@ -1,3 +1,7 @@
...
 PATH_PREFIX=/v1/status
+<<<<<<< HEAD
```
Comment (humphd) [inline src/api/status/env.local:3]:
```
git checkout master src/api/status/env.local
```

File: src/api/status/src/server.js (hunk 2 above)
Comment (humphd) [reply src/api/status/src/server.js:46]:
```
git checkout master src/api/status/src/server.js
```

File: src/web/src/components/SearchInput/SearchInput.tsx [hunk 7]
```diff
@@ -5,7 +5,7 @@ import useSearchValue from '../../hooks/use-search-value';
 const SearchInput = () => {
   const { filter } = useSearchValue();
 
-  return filter === 'author' ? <AuthorSearchInput /> : <PostSearchInput />;
+  return <>{filter === 'author' ? <AuthorSearchInput /> : <PostSearchInput />}</>;
```
Comment (humphd) [inline src/web/src/components/SearchInput/SearchInput.tsx:8]: Can you help me understand more about this change?  What does wrapping in an element do for us in this case that returning the component won't?

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/web/src/components/SearchInput/SearchInput.tsx (hunk 7 above)
Comment (DukeManh) [reply src/web/src/components/SearchInput/SearchInput.tsx:8]: @humphd, `filter === 'author' ? <AuthorSearchInput /> : <PostSearchInput />;` is like an function that's called in every re-render, and its return result in unknown. When either `AuthorSearchInput` or `PostSearchInput` is returned, the current input is unmounted and re-rendered so the input loses focus without `autofoucs` attribute.


`<>{filter === 'author' ? <AuthorSearchInput /> : <PostSearchInput />}</>` is a React Component that's has conditional rendering, `AuthorSearchInput` and `PostSearchInput` won't re-mount if it doesn't change.

I believe that's how React works, let me tag Dan Abramov to confirm :).

**Comment by DukeManh** [comment]
@drew5494 , can you rebase to pickup the new changes?

File: src/web/src/components/SearchInput/SearchInput.tsx (hunk 7 above)
Comment (humphd) [reply src/web/src/components/SearchInput/SearchInput.tsx:8]: In both cases the React component tree will be different, whether at the root or child level. "filter === 'author' ? <AuthorSearchInput /> : <PostSearchInput />; is like an function" I don't think so.  It's the same expression you're evaluating in `<>{filter === 'author' ? <AuthorSearchInput /> : <PostSearchInput />}</>`

```js
function ComponentA() { return <div></div>; }
function ComponentB() { return <div></div>; }

const x = 1;

const element = <>{x === 1 ? <ComponentA /> : <ComponentB />}</>;
// var element = /*#__PURE__*/React.createElement(React.Fragment, null, x === 1 ? /*#__PURE__*/React.createElement(ComponentA, null) : /*#__PURE__*/React.createElement(ComponentB, null));

const element2 = x === 1 ? <ComponentA /> : <ComponentB />;
// var element2 = x === 1 ? /*#__PURE__*/React.createElement(ComponentA, null) : /*#__PURE__*/React.createElement(ComponentB, null);
```

If there's a real difference when wrapping in a fragment, related to re-rendering, I'd like to understand what it is. Both are going to evaluate to, and produce the same DOM.

Comment (humphd) [reply src/web/src/components/SearchInput/SearchInput.tsx:8]: Are you saying that the fragment provides stability between re-renders in order to maintain focus?

Comment (DukeManh) [reply src/web/src/components/SearchInput/SearchInput.tsx:8]: > Are you saying that the fragment provides stability between re-renders in order to maintain focus?

Yes, I mean that.
The difference is that the React fragment prevents unmounting, both will produce the same DOM, but the first one doesn't re-mount and the second one does on every re-render.

Comment (humphd) [reply src/web/src/components/SearchInput/SearchInput.tsx:8]: OK, that makes sense.

**Comment by drew5494** [comment]
> @drew5494 , can you rebase to pickup the new changes?

Sure. Once I have a chance to figure this out.

**Comment by DukeManh** [comment]
@drew5494  
```
git fetch upstream master
git rebase upstream/master
```

**Comment by TueNguyen2911** [comment]
Looks like you still have 2 commits, rebase is not correct

**Comment by DukeManh** [comment]
@TueNguyen2911 it's ok, don't have to squash during rebase, we can squash when this is merged.

**Comment by TueNguyen2911** [comment]
> @TueNguyen2911 it's ok, don't have to squash during rebase, we can squash when this is merged.

Aah, that's good then

**Review by humphd — DISMISSED** [review]
This is a nice fix.  Works great on Vercel.

**Review by DukeManh — DISMISSED** [review]

**Comment by humphd** [comment]
@drew5494 can you squash and rebase this for me so I can merge?

**Comment by DukeManh** [comment]
@drew5494, your branch is still out-of-date, I wonder if you have the right remote URL.
```
git remote -v
git remote set-url upstream https://github.com/Seneca-CDOT/telescope.git
git fetch upstream master
git rebase upstream/master
```

**Comment by humphd** [comment]
Rebased via GitHub.
